NLog.Web.AspNetCore/LayoutRenderers/AspNetRequestPostedBody.cs

```diff
@@ -47,75 +50,108 @@ protected override void DoAppend(StringBuilder builder, LogEventInfo logEvent)
+        }
 
-            if (!body.CanRead)
-            {
-                InternalLogger.Debug("AspNetRequestPostedBody: body stream has been closed");
-                return;
-            }
+        private static string BodyToString(Stream body)
+        {
+            //note: dispose of StreamReader isn't doing things besides closing the stream (which can be turn off, and then it's a NOOP)
+            var bodyReader = new StreamReader(body);
+            var content = bodyReader.ReadToEnd();
+            return content;
+        }
 
-            // reset if possible
-            long oldPosition = body.Position;
+        private bool TryResetStream(HttpRequest httpRequest, out long oldPosition, ref Stream body)
+        {
+            long? contentLength = httpRequest.ContentLength;
+            oldPosition = body.Position;
             if (!body.CanSeek)
             {
                 if (oldPosition > 0 && oldPosition >= contentLength)
                 {
                     InternalLogger.Debug("AspNetRequestPostedBody: body stream cannot seek and already read. StreamPosition={0}", oldPosition);
-                    return;
+                    return true;
                 }
 
                 oldPosition = 0;
                 if (!TryEnableBuffering(httpRequest, contentLength, out body))
                 {
-                    return; // Body cannot be buffered
+                    return true;
                 }
             }
             else
             {
                 if (MaxContentLength > 0 && !contentLength.HasValue && body.Length > MaxContentLength)
                 {
                     InternalLogger.Debug("AspNetRequestPostedBody: body stream too big. Body.Length={0}", body.Length);
-                    return; // Body too large
+                    return true;
                 }
             }
 
             body.Position = 0;
+            return false;
+        }
 
-            //note: dispose of StreamReader isn't doing things besides closing the stream (which can be turn off, and then it's a NOOP)
-            var bodyReader = new StreamReader(body);
-            var content = bodyReader.ReadToEnd();
+        private bool TryGetBody(HttpRequest httpRequest, out Stream body)
+        {
+            long? contentLength = httpRequest.ContentLength;
+            body = null;
+            if (contentLength <= 0)
+            {
+                return true;
+            }
 
-            //restore
-            body.Position = oldPosition;
+            if (MaxContentLength > 0 && contentLength > MaxContentLength)
+            {
+                InternalLogger.Debug("AspNetRequestPostedBody: body stream is too big. ContentLength={0}", contentLength);
+                return true;
+            }
 
-            builder.Append(content);
+body = GetBodyStream(httpRequest);
+
+if (body == null)
+            {
+                InternalLogger.Debug("AspNetRequestPostedBody: body stream was null");
+                return true;
```

Comment:
Why not false?

Reply:
good one. I will inverse it (this was the choice of R# ;)